genie-agent/src/main/java/com/netflix/genie/agent/execution/services/LaunchJobServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+ *
+ */
+
+package com.netflix.genie.agent.execution.services;
+
+import com.netflix.genie.agent.execution.ExecutionContext;
+import com.netflix.genie.agent.execution.exceptions.JobLaunchException;
+import com.netflix.genie.agent.utils.PathUtils;
+import com.netflix.genie.common.dto.v4.JobSpecification;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Configures and launches a job sub-process using metadata passed through ExecutionContext.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+@Component
+@Lazy
+class LaunchJobServiceImpl implements LaunchJobService {
+
+    private final ExecutionContext executionContext;
+
+    LaunchJobServiceImpl(final ExecutionContext executionContext) {
+        this.executionContext = executionContext;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Process launchProcess() throws JobLaunchException {
+        final JobSpecification jobSpec = executionContext.getJobSpecification();
+
+        if (jobSpec == null) {
+            throw new JobLaunchException("Job specification is not set in execution context");
+        }
+
+        final ProcessBuilder processBuilder = new ProcessBuilder();
+
+        configureEnvironment(processBuilder);
+        configureCommandArguments(processBuilder, jobSpec);
+        configureIO(processBuilder, jobSpec);
+
+        try {
+            return processBuilder.start();
+        } catch (final IOException | SecurityException e) {
+            throw new JobLaunchException("Failed to launch job: ", e);
+        }
+    }
+
+    private void configureEnvironment(
+        final ProcessBuilder processBuilder
+    ) throws JobLaunchException {
+
+        final File jobDirectory = executionContext.getJobDirectory();
+
+        // Validate job running directory
+        if (jobDirectory == null) {
+            throw new JobLaunchException("Job directory is not set in execution context");
+        } else if (!jobDirectory.exists()) {
+            throw new JobLaunchException("Job directory does not exist: " + jobDirectory);
+        } else if (!jobDirectory.isDirectory()) {
+            throw new JobLaunchException("Job directory is not a directory: " + jobDirectory);
+        } else if (!jobDirectory.canWrite()) {
+            throw new JobLaunchException("Job directory is not writeable: " + jobDirectory);
+        }
+
+        // Configure job running directory
+        processBuilder.directory(jobDirectory);
+
+        final Map<String, String> currentEnvironmentVariables = processBuilder.environment();
+        final Map<String, String> jobEnvironmentVariables = executionContext.getJobEnvironment();
+
+        // Validate additional environment variables
+        if (jobEnvironmentVariables == null) {
```

Comment:
Prefer Optional

Reply:
Same as above, shouldn't be possible to get here with this not set.
(and if we ever do, this is a nicer error than an NPE)